From 9a8dee3dae29b464a42edd83a66226c313e25ea2 Mon Sep 17 00:00:00 2001
From: Pratham Patel <ppatel@ciq.com>
Date: Sun, 27 Jul 2025 19:18:55 +0530
Subject: [PATCH] ceph: avoid putting the realm twice when decoding snaps fails

jira VULN-65927
cve CVE-2022-49770
commit-author Xiubo Li <xiubli@redhat.com>
commit 51884d153f7ec85e18d607b2467820a90e0f4359
upstream-diff Absence of the following commit caused a merge conflict:
    2e586641c950 ("ceph: do not update snapshot context when there is no new snapshot")

When decoding the snaps fails it maybe leaving the 'first_realm'
and 'realm' pointing to the same snaprealm memory. And then it'll
put it twice and could cause random use-after-free, BUG_ON, etc
issues.

	Cc: stable@vger.kernel.org
Link: https://tracker.ceph.com/issues/57686
	Signed-off-by: Xiubo Li <xiubli@redhat.com>
	Reviewed-by: Ilya Dryomov <idryomov@gmail.com>
	Signed-off-by: Ilya Dryomov <idryomov@gmail.com>
(cherry picked from commit 51884d153f7ec85e18d607b2467820a90e0f4359)
	Signed-off-by: Pratham Patel <ppatel@ciq.com>
---
 fs/ceph/snap.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/fs/ceph/snap.c b/fs/ceph/snap.c
index 3360a131490f..1354bc1d7f3a 100644
--- a/fs/ceph/snap.c
+++ b/fs/ceph/snap.c
@@ -665,7 +665,7 @@ int ceph_update_snap_trace(struct ceph_mds_client *mdsc,
 	struct ceph_mds_snap_realm *ri;    /* encoded */
 	__le64 *snaps;                     /* encoded */
 	__le64 *prior_parent_snaps;        /* encoded */
-	struct ceph_snap_realm *realm = NULL;
+	struct ceph_snap_realm *realm;
 	struct ceph_snap_realm *first_realm = NULL;
 	int invalidate = 0;
 	int err = -ENOMEM;
@@ -673,6 +673,7 @@ int ceph_update_snap_trace(struct ceph_mds_client *mdsc,
 
 	dout("update_snap_trace deletion=%d\n", deletion);
 more:
+	realm = NULL;
 	ceph_decode_need(&p, e, sizeof(*ri), bad);
 	ri = p;
 	p += sizeof(*ri);
-- 
2.47.3

